simulator/kernel/mars/t/unit_test.cpp

```diff
@@ -236,59 +236,69 @@ TEST_CASE( "MARS: open non existing file")
     CHECK( sys.sim->read_cpu_register( v0) == all_ones<uint64>());
 }
 
-TEST_CASE( "MARS: open, write, read and close a file")
+static void write_buff_to_file( System* sys, uint64 descr,
+                                uint64 buff_ptr, uint64 chars_to_write)
+{
+    sys->sim->write_cpu_register( v0, 15u); //write to file
+    sys->sim->write_cpu_register( a0, descr);
+    sys->sim->write_cpu_register( a1, buff_ptr);
+    sys->sim->write_cpu_register( a2, chars_to_write);
+}
+
+static void open_file( System* sys, uint64 filename_ptr, uint64 flags)
+{
+    sys->sim->write_cpu_register( v0, 13u); // open file
+    sys->sim->write_cpu_register( a0, filename_ptr);
+    sys->sim->write_cpu_register( a1, flags);
+    //return sys->sim->read_cpu_register( v0); //FIXME: tests failed
+}
+
+static void read_from_file( System* sys, uint64 descr,
+                            uint64 buff_ptr, uint64 chars_to_read)
+{
+    sys->sim->write_cpu_register( v0, 14u); // read from file
+    sys->sim->write_cpu_register( a0, descr);
+    sys->sim->write_cpu_register( a1, buff_ptr);
+    sys->sim->write_cpu_register( a2, chars_to_read);
+}
+
+TEST_CASE( "MARS: open, read and close a file")
 {
     std::string filename("tempfile");
     std::ostringstream output;
     System sys( std::cin, output);
     sys.mem->write_string( filename, 0x1000);
     sys.mem->write_string( "Lorem Ipsum\n", 0x2000);
 
-    sys.sim->write_cpu_register( v0, 13u); // open file
-    sys.sim->write_cpu_register( a0, 0x1000u);
-    sys.sim->write_cpu_register( a1, 1); // write
+    open_file(&sys, 0x1000, 1); // open WRONLY
+    //auto descriptor = sys.sim->read_cpu_register( v0);
+    //FIXME test fails if it's used in this sequence
     CHECK( sys.mars_kernel->execute() == Trap::NO_TRAP);
     auto descriptor = sys.sim->read_cpu_register( v0);
     CHECK( descriptor >= 3);
-
-    sys.sim->write_cpu_register( v0, 15u); // write to file
-    sys.sim->write_cpu_register( a0, descriptor);
-    sys.sim->write_cpu_register( a1, 0x2000);
-    sys.sim->write_cpu_register( a2, 11);
+    write_buff_to_file(&sys, descriptor, 0x2000, 11); // write string to file
```

Comment:
Good unit tests should have one Arrangement, one Act and one Assertion section, and the problem here is that Act and Assertion phases are mixed. In other words, we test that open was correct, than we test that write was succsessful, than we check the data read etc. As a result, if the test fails, we won't be able to say what stage failed without examining the unit test source code.